### ISSUE TITLE
Weird mod loading issue

### ISSUE THREAD

**Issue by milochristiansen** [anegostudios/VintageStory-Issues#12]
```
22:47:55 [Server Error] Failed loading a mod :0 CS0006 Metadata file `VintagestoryAPI.dll' could not be found
:0 CS0006 Metadata file `Newtonsoft.Json.dll' could not be found
:0 CS0006 Metadata file `protobuf-net.dll' could not be found
:0 CS0006 Metadata file `Tavis.JsonPatch.dll' could not be found

22:47:55 [Server Warning] Mod Set BrickInTheWall.zip could not be loaded and will be ignored!
```

All the other mods loaded fine, but IIRC this is the only one that uses loose `.cs` files.

The server is started via `mono /home/vintage/server/VintagestoryServer.exe --dataPath /home/vintage/data/gamedata` from the `/home/vintage` directory.

This is on Ubuntu 16.04.3 LTS, Mono 4.6.2.7 with Vintage Story 1.5.2.1

**Comment by copygirl** [comment]
Is this only an issue on the server or does the mod also not load when you're playing singleplayer?

**Comment by milochristiansen** [comment]
It loads fine in single player. I'm guessing it has something to do with the working directory not being the binary directory, but I'm not familiar enough with .NET to tell for sure.

I don't really want to experiment too much on the server, since I don't want to screw anything up.

**Comment by tyronx** [comment]
https://stackoverflow.com/questions/14490247/csc-error-cs0006-metadata-file-could-not-be-found-building-a-net-3-5-applicatio mentions C# 7 language features as issue - which would make sense because mono 4.6 does not support only C# 6.

Assuming to be this issue. Please reopen if that is not the case.

**Comment by milochristiansen** [comment]
I added `<LangVersion>6</LangVersion>` to the .csproj file and compiled, and it worked fine, so that shouldn't be the issue here. (To make sure that key was working correctly I set it to various other values to see what errors I got, for example it won't compile with `iso-1`.)

**Comment by tyronx** [comment]
Send me source files please

**Comment by milochristiansen** [comment]
[BrickTransitionBehavior.zip](https://github.com/anegostudios/VintageStory-Issues/files/1830601/BrickTransitionBehavior.zip)

**Comment by milochristiansen** [comment]
The whole mod is [over here](https://www.vintagestory.at/forums/topic/407-brick-in-the-wall-seamless-brick-transition-mod/)

**Comment by tyronx** [comment]
Did you run this from within the IDE or from console?

**Comment by milochristiansen** [comment]
When test compiling I used the console (`dotnet build`), when running on the server I used the zip linked above with the command in the first post (the zip was in `/home/vintage/data/gamedata/Mods`).

**Comment by milochristiansen** [comment]
Just to be 100% clear, when I compiled the .cs file and tried to load the mod with a dll instead of the .cs it worked. I would prefer not to have to do this, but for now that will be an acceptable workaround.

Whatever the issue is, it is limited to loading raw .cs files, probably only in mono.

**Comment by tyronx** [comment]
Thanks for the feedback. Another theory is that the built-in CSharpCodeProvider seemingly only compiles c# code until C# 5.0 and thus causes issues when trying to reference C# 6.0 libs.

**Comment by tyronx** [comment]
Hm but then again, my source compiled mod works just fine on debian / mono 5.4.1.7
What if you update to mono 5?

**Comment by milochristiansen** [comment]
Sorry, I read this yesterday, but forgot to reply :(

The system in question isn't mine, so I don't want to mess with it any more than I have to.

**Comment by tyronx** [comment]
@copygirl does your updates on the modloader also load stuff in order of dependency? This should fix this problem then.

**Comment by copygirl** [comment]
I don't think I do ordering, but it's something that could be done.

**Comment by tyronx** [comment]
@milochristiansen does this issue still happen with 1.5.3?

**Comment by tyronx** [comment]
copygirls refactorings should fix this, so assuming this is fixed. Please reopen otherwise

**Comment by milochristiansen** [comment]
This is *not* fixed.

```
3:22:20 PM [Server Warning]: Error discovering files for 'BrickInTheWall.zip': 'BrickTransitionBehavior.cs' found at invalid path 'BrickTransitionBehavior.cs'. All .cs files should be in a subfolder 'src' of the mod folder / archive. This will be an ERROR in the future!
3:22:20 PM [Server Warning]: Mod 'BrickInTheWall.zip' (brickinthewall): Error parsing version string '1.2' at index 3: Expected PATCH version, found end of string (best guess: 1.2.0)
3:22:20 PM [Server Error]: Compiler errors during compiling mod 'BrickInTheWall.zip' (brickinthewall):
3:22:20 PM [Server Error]: [:0] [CS0006] Metadata file `VintagestoryAPI.dll' could not be found
3:22:20 PM [Server Error]: [:0] [CS0006] Metadata file `Lib/Newtonsoft.Json.dll' could not be found
3:22:20 PM [Server Error]: [:0] [CS0006] Metadata file `Lib/protobuf-net.dll' could not be found
3:22:20 PM [Server Error]: [:0] [CS0006] Metadata file `Lib/Tavis.JsonPatch.dll' could not be found
3:22:20 PM [Server Error]: Error loading mod 'BrickInTheWall.zip' (brickinthewall):
3:22:20 PM [Server Error]: System.Exception: Error(s) compiling mod 'BrickInTheWall.zip' (brickinthewall) from source files
3:22:20 PM [Server Error]: at _gpHtMiskGLjqPGZxwTher6nxdoG._KR6aEbe5eiNuqwaa2EH6wfBLvSd (_QtTCbWTgLzQ0L5pnJ4SA0p0S4nW , System.Collections.Generic.IEnumerable`1[T] ) [0x00141] in <00cbb88cfc944a46ba4ac8d0bbad9be6>:0
3:22:20 PM [Server Error]: at _QtTCbWTgLzQ0L5pnJ4SA0p0S4nW._QsEdKSeIRDVeKF6yvYWCoIe6ZXp (Vintagestory.API.Common.ILogger , _gpHtMiskGLjqPGZxwTher6nxdoG , _qTMXzQjTmZCapFMQ0QaS7GReooq ) [0x000d3] in <00cbb88cfc944a46ba4ac8d0bbad9be6>:0
```

**Comment by copygirl** [comment]
What happens if you change to the directory the `VintagestoryServer.exe` is in before you run it? That would confirm that the working directory is the issue.

**Comment by milochristiansen** [comment]
Well, since the server exhibiting this issue is being run under the monitor, and it would be a pain in the rear to run it manually, I decided to see what would happen if I ran the VS server with the current working directory *not* set to the same directory as the EXE on my PC. The result? I can reproduce the issue on my system now.

It seems this is 100% reliable. Run VS with the working directory set to anything other than the binary directory, and it fails to load mods that are not compiled.

BTW: This issue is still marked closed, and I cannot change that.
